devito/equation.py

```diff
@@ -188,4 +201,6 @@ def solve(eq, target, **kwargs):
     # turnaround time
     kwargs['rational'] = False  # Avoid float indices
     kwargs['simplify'] = False  # Do not attempt premature optimisation
-    return sympy.solve(eq, target, **kwargs)[0]
+    if hasattr(eq, 'lhs'):
```

Comment:
???

Reply:
Just turn input into `expr=0` so if it's an equation do diff lhs-rhs. This way don't pass eq to solve. Solve(eq) returns sympy.add and we back to wrong types

Reply:
cfd tutorials 5 and 6 fail without this.

Reply:
what I'm trying to say is that ``hasattr`` should be avoided at all costs, when possible. Here it seems possible: all and only ``Eq`` have a ``lhs`` attribute, so this should rather be
```
if isinstance(eq, Expr):  # in fact, change the var name from eq into expr
    eq = eq.lhs - eq.rhs
```

FWIW, I also tried dropping the ``if`` entirely (ie, even without the tweak above) and both the ``cfd`` and all tests in ``test_derivative.py`` pass

Despite this might be seen as nitpicking, it really is not, because it's self-documenting what we are prepared to accept in ``solve`` (which, one day, might become our own (faster) ``solve``)

Reply:
>FWIW, I also tried dropping the if entirely (ie, even without the tweak above) and both the cfd and all tests in test_derivative.py pass

I can guaranty you it does not pass. Can change to isinstance if you want fine with me

Reply:
no I mean I was dropping the ``if``, but I still had the ``eq = eq.lhs - eq.rhs``

yes, turning into ``isinstance(expr, Eq)`` would be preferable 

do you get why it's better this way? basically because it's a stronger (less loose) check, and this is a fundamental piece of API (as you've often rightly pointed out)

Reply:
Changed. Yes I get it, it's more robust, was tryign to avoid extra imports but that's stupid Eq already imported, also wondering what will happen if an `Inc` is passed now but for later. Anyway I think dropping `if` is dangerous so I kept it

Reply:
maybe we should raise a ValueError if an Inc ? anyway...that's fine for now